File: src/13-challenge/solution.py
from typing import Literal, Callable, Dict, Final, cast


Cell = Literal[">", "<", "^", "v", "."]
Result = Literal["completed", "broken", "loop"]

Movements = Dict[Cell, Callable[[], None]]


def fn(factory: list[str]) -> Result:
    gift: list[int] = [0, 0]
    gift_path: set[str] = set()

    def move_right():
        gift[1] += 1

    def move_left():
        gift[1] -= 1

    def move_up():
        gift[0] -= 1

    def move_down():
        gift[0] += 1

    movements: Final[Movements] = {
        ">": move_right,
        "<": move_left,
        "^": move_up,
        "v": move_down,
    }

    rows: Final = len(factory)
    cols: Final = len(factory[0]) if factory[0] else 0

    while gift[0] > -1 and gift[0] < rows and gift[1] > -1 and gift[1] < cols:
        movement = cast(Cell, factory[gift[0]][gift[1]])
        if movement == ".":
            return "completed"

        gift_path.add(f"[{gift[0]}, {gift[1]}]")

        movements[movement]()
        if f"[{gift[0]}, {gift[1]}]" in gift_path:
            return "loop"

    return "broken"

File: src/13-challenge/solution.js
/**
 * @typedef {'>' | '<' | '^' | 'v' | '.'} Cell
 */

/**
 * @param {string[]} factory - The factory layout
 * @returns {'completed'|'broken'|'loop'} Result of the gift journey
 */
export default function fn(factory) {
	/** @type {[number, number]} */
	const gift = [0, 0]

	/** @type {const giftPath: } */
	const giftPath = new Set()

	/** @type {Record<Exclude<Cell, '.'>, () => number>} */
	const movements = {
		'>': () => gift[1]++,
		'<': () => gift[1]--,
		'^': () => gift[0]--,
		'v': () => gift[0]++,
	}

	const rows = factory.length
	const cols = (factory[0] ?? '').length

	while (gift[0] > -1 && gift[0] < rows && gift[1] > -1 && gift[1] < cols) {
		const /** @type {Cell} */ movement = factory[gift[0]][gift[1]]
		if (movement === '.') return 'completed'

		giftPath.add(`[${gift[0]}, ${gift[1]}]`)

		movements[movement]()
		if (giftPath.has(`[${gift[0]}, ${gift[1]}]`)) return 'loop'
	}

	return 'broken'
}
